plugin/plugin.go

```diff
@@ -0,0 +1,369 @@
+		}
+		for varName, varVal := range p.SysVars {
+			if oldPlugin.SysVars[varName] == nil || *oldPlugin.SysVars[varName] != *varVal {
+				return errors.Errorf("reload plugin with different sysVar is unsupported %v", varVal)
+			}
+		}
+	}
+	if p.RequireVersion != nil {
+		for component, reqVer := range p.RequireVersion {
+			if ver, ok := tiPlugins.versions[component]; !ok || ver < reqVer {
+				return errors.Errorf("plugin %s require %s be %v but got %v", p.Name, component, reqVer, ver)
+			}
+		}
+	}
+	if p.SysVars != nil {
+		for varName := range p.SysVars {
+			if !strings.HasPrefix(varName, p.Name) {
+				return errors.Errorf("plugin %s's sysVar %s must be prefix with plugin name %s", p.Name, varName, p.Name)
+			}
+		}
+	}
+	if p.Manifest.Validate != nil {
+		if err := p.Manifest.Validate(ctx, p.Manifest); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Init initializes the plugin and load plugin by config param.
+// This method isn't thread-safe and must be called before any other plugin operation.
+func Init(ctx context.Context, cfg Config) (err error) {
+	tiPlugins := &plugins{
+		plugins:      make(map[Kind][]Plugin),
+		versions:     make(map[string]uint16),
+		dyingPlugins: make([]Plugin, 0),
+	}
+
+	// setup component version info for plugin running env.
+	for component, version := range cfg.EnvVersion {
+		tiPlugins.versions[component] = version
+	}
+
+	// load plugin dl & manifest.
+	for _, pluginID := range cfg.Plugins {
+		var pName string
+		pName, _, err = ID(pluginID).Decode()
+		if err != nil {
+			return
+		}
+		// check duplicate.
+		_, dup := tiPlugins.versions[pName]
+		if dup {
+			if cfg.SkipWhenFail {
+				continue
+			}
+			err = errors.Errorf("plugin [%s] be duplicate declared in init parameters", pluginID)
+			return
+		}
+		// load dl.
+		var plugin Plugin
+		plugin, err = loadOne(cfg.PluginDir, ID(pluginID))
+		if err != nil {
+			if cfg.SkipWhenFail {
+				continue
+			}
+			return
+		}
+		tiPlugins.add(plugin)
+	}
+
+	// cross validate & load plugins.
+	for kind := range tiPlugins.plugins {
+		for i := range tiPlugins.plugins[kind] {
+			if err = tiPlugins.plugins[kind][i].validate(ctx, tiPlugins, initMode); err != nil {
+				if cfg.SkipWhenFail {
+					tiPlugins.plugins[kind][i].State = Disable
+					err = nil
+					continue
+				}
+				return
+			}
+			p := tiPlugins.plugins[kind][i]
+			if err = p.OnInit(ctx, p.Manifest); err != nil {
+				if cfg.SkipWhenFail {
+					tiPlugins.plugins[kind][i].State = Disable
+					err = nil
+					continue
+				}
+				return
+			}
+			if cfg.GlobalSysVar != nil {
+				for key, value := range tiPlugins.plugins[kind][i].SysVars {
+					(*cfg.GlobalSysVar)[key] = value
+					if value.Scope != variable.ScopeSession && cfg.PluginVarNames != nil {
+						*cfg.PluginVarNames = append(*cfg.PluginVarNames, key)
+					}
+				}
+			}
+			tiPlugins.plugins[kind][i].State = Ready
+		}
+	}
+	pluginGlobal = copyOnWriteContext{tiPlugins: unsafe.Pointer(tiPlugins)}
```

Comment:
Why not just use a simple RWMutex on pluginGlobal ?

Reply:
pluginGlobal's change is infrequent, "copy on write" seem to be nature way